<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Examples · GenomicAnnotations.jl</title><meta name="title" content="Examples · GenomicAnnotations.jl"/><meta property="og:title" content="Examples · GenomicAnnotations.jl"/><meta property="twitter:title" content="Examples · GenomicAnnotations.jl"/><meta name="description" content="Documentation for GenomicAnnotations.jl."/><meta property="og:description" content="Documentation for GenomicAnnotations.jl."/><meta property="twitter:description" content="Documentation for GenomicAnnotations.jl."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">GenomicAnnotations.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">GenomicAnnotations.jl</a></li><li><a class="tocitem" href="../io/">I/O</a></li><li><a class="tocitem" href="../accessing/">Accessing and modifying annotations</a></li><li><a class="tocitem" href="../loci/">Representing genomic loci</a></li><li><a class="tocitem" href="../genes/">Filtering: the @genes macro</a></li><li class="is-active"><a class="tocitem" href>Examples</a><ul class="internal"><li><a class="tocitem" href="#Adding-chromosome-name-to-all-locus-tags"><span>Adding chromosome name to all locus tags</span></a></li><li><a class="tocitem" href="#Adding-qualifiers"><span>Adding qualifiers</span></a></li><li><a class="tocitem" href="#Converting-between-formats"><span>Converting between formats</span></a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Examples</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Examples</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/BioJulia/GenomicAnnotations.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/BioJulia/GenomicAnnotations.jl/blob/master/docs/src/examples.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Examples"><a class="docs-heading-anchor" href="#Examples">Examples</a><a id="Examples-1"></a><a class="docs-heading-anchor-permalink" href="#Examples" title="Permalink"></a></h1><h2 id="Adding-chromosome-name-to-all-locus-tags"><a class="docs-heading-anchor" href="#Adding-chromosome-name-to-all-locus-tags">Adding chromosome name to all locus tags</a><a id="Adding-chromosome-name-to-all-locus-tags-1"></a><a class="docs-heading-anchor-permalink" href="#Adding-chromosome-name-to-all-locus-tags" title="Permalink"></a></h2><p>When iterating over genes, the parent chromosome can be accessed with <code>parent(::Gene)</code>.</p><pre><code class="language-julia hljs">using GenomicAnnotations
chrs = readgbk(&quot;genome.gbk&quot;)
for gene in @genes(chrs)
    gene.locus_tag = string(parent(gene).name, &quot;_&quot;, gene.locus_tag)
end
printgbk(&quot;updated_genome.gbk&quot;, chrs)</code></pre><h2 id="Adding-qualifiers"><a class="docs-heading-anchor" href="#Adding-qualifiers">Adding qualifiers</a><a id="Adding-qualifiers-1"></a><a class="docs-heading-anchor-permalink" href="#Adding-qualifiers" title="Permalink"></a></h2><p>GenomicAnnotations supports arbitrary qualifiers, so you can add any kind of information. The following script reads and adds the output from <a href="http://phobius.sbc.su.se/">Phobius</a> (a predictor for transmembrane helices) to the annotations.</p><pre><code class="language-julia hljs">using GenomicAnnotations
chrs = readgbk(&quot;genome.gbk&quot;)

function addphobius!(chr, file)
    @progress for line in readlines(file)
        m = match(r&quot;^(\w+) +(\d+) +&quot;, line)
        if m != nothing
            locus_tag = m[1]
            tmds = parse(Int, m[2])
            @genes(chr, CDS, :locus_tag == locus_tag).phobius .= tmds
        end
    end
end

addphobius!(chrs, &quot;phobius.txt&quot;)

open(GenBank.Writer, &quot;updated_genome.gbk&quot;) do w
    for chr in chrs
        write(w, chr)
    end
end</code></pre><h2 id="Converting-between-formats"><a class="docs-heading-anchor" href="#Converting-between-formats">Converting between formats</a><a id="Converting-between-formats-1"></a><a class="docs-heading-anchor-permalink" href="#Converting-between-formats" title="Permalink"></a></h2><p>Annotations can be read from one file format and written as another. Converting between the supported human-readable formats (GenBank and EMBL) or the tab-delimited formats (GFF3 and GTF) will likely work out of the box, but converting from a human-readable format to a tab-delimited format, or vice versa, may need some human intervention. Currently, GenomicAnnotations does not make any attempt to rename columns or perform any sanity checks to ensure that the resulting file meets specifications. Refer to the respective format specifications for details on what attributes need to be included, etc. Notably, GenBank and GFF3 headers do not contain the same information, and GTF files lack a header altogether, thus all information in the header is lost when saving annotations as another format. GTF files also do not allow the inclusion of sequence data, unlike GFF3. Below is a simple example script that demonstrates some of the changes that need to be made. If your use-case includes more complex features, such as multi-exon genes, you will likely need to make more changes as part of the convertion.</p><pre><code class="language-julia hljs">using GenomicAnnotations
using DataFrames
chrs = readgbk(&quot;genome.gbk&quot;)
open(GFF.Writer, &quot;genome.gff&quot;) do w
    for chr in chrs
        # GenBank features often contain features that are not usually included
        # in GFF3 files, so let&#39;s remove some:
        cols_to_remove = intersect([&quot;translation&quot;, &quot;mol_type&quot;, &quot;organism&quot;], names(chr.genedata))
        if !isempty(cols_to_remove)
            chr.genedata = chr.genedata[:, Not(cols_to_remove)]
        end
        # The GenBank format uses the :source feature to store metadata about
        # the record, but in GFF3, :source is the name of the column which
        # contains the sequence name. Thus, we need to change the GenBank
        # :source to the GFF3 equivalent :region. According to the GenBank
        # specification, :source is mandatory, but it&#39;s best to be safe and
        # check that it&#39;s really there:
        source_entries = @genes(chr, source)
        if !isempty(source_entries)
            for source in source_entries
                region = feature!(source, :region)
                region.Name = chr.name
                region.ID = chr.name
                # GenBank files include information about circularity of a
                # contig in its header, but in the GFF3 format this
                # information is encoded in the &quot;Is_circular&quot; attribute of
                # the first :region feature:
                if occursin(&quot;circular&quot;, chr.header)
                    region.Is_circular = true
                end
            end
        else
            # If the :source feature is missing, we&#39;ll have to create a :region
            # from scratch:
            addgene!(chr, &quot;region&quot;, 1:length(chr.sequence);
                Name = chr.name,
                ID = chr.name,
                Is_circular = occursin(&quot;circular&quot;, chr.header))
            sort!(chr.genes)
        end
        # Most features, such as :CDS or :tRNA, have a corresponding :gene
        # that it belongs to. In GFF3, this hierarchical relationship is shown
        # using the &quot;ID&quot; and &quot;Parent&quot; attributes. Here, we set the &quot;ID&quot;
        # attribute of all :gene features to match their &quot;locus_tag&quot;, and then
        # set the &quot;Parent&quot; attributes of all non-:gene features to match the
        # &quot;ID&quot; of their respective :gene, if there is one:
        gene_features = @genes(chr, gene)
        gene_features.ID .= gene_features.locus_tag
        for gene in @genes(chr, !gene)
            if get(gene, :locus_tag, &quot;missing&quot;) in skipmissing(gene_features.locus_tag)
                gene.Parent = gene.locus_tag
            end
        end
        write(w, chr)
    end
end</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../genes/">« Filtering: the @genes macro</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Thursday 21 November 2024 16:05">Thursday 21 November 2024</span>. Using Julia version 1.11.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
