<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>GenomicAnnotations.jl · GenomicAnnotations.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link href="assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>GenomicAnnotations.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li class="current"><a class="toctext" href>GenomicAnnotations.jl</a><ul class="internal"><li><a class="toctext" href="#Description-1">Description</a></li><li><a class="toctext" href="#Installation-1">Installation</a></li><li><a class="toctext" href="#Examples-1">Examples</a></li></ul></li><li><a class="toctext" href="io/">I/O</a></li><li><a class="toctext" href="accessing/">Accessing and modifying annotations</a></li><li><a class="toctext" href="genes/">Filtering: the @genes macro</a></li></ul></nav><article id="docs"><header><nav><ul><li><a href>GenomicAnnotations.jl</a></li></ul><a class="edit-page" href="https://github.com/kdyrhage/GenomicAnnotations.jl/blob/master/docs/src/index.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>GenomicAnnotations.jl</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="GenomicAnnotations.jl-1" href="#GenomicAnnotations.jl-1">GenomicAnnotations.jl</a></h1><h2><a class="nav-anchor" id="Description-1" href="#Description-1">Description</a></h2><p>GenomicAnnotations is a package for reading, modifying, and writing genomic annotations in the GenBank format.</p><h2><a class="nav-anchor" id="Installation-1" href="#Installation-1">Installation</a></h2><pre><code class="language-julia">julia&gt;]
pkg&gt; add GenomicAnnotations</code></pre><p>or</p><pre><code class="language-julia">using Pkg
Pkg.add(&quot;GenomicAnnotations&quot;)</code></pre><h2><a class="nav-anchor" id="Examples-1" href="#Examples-1">Examples</a></h2><p>GenBank files are read with <code>readgbk(pathtofile)</code>, which returns a vector of <code>Chromosome</code>s. <code>gbkfile</code> can be gzipped as long as the filename ends in &quot;.gz&quot;. If we&#39;re only interested in the first chromosome in <code>example.gbk</code> we only need to store the first element.</p><pre><code class="language-julia">chr = readgbk(&quot;test/example.gbk&quot;)[1]</code></pre><p><code>Chromosome</code>s have five fields, <code>name</code>, <code>header</code>, <code>genes</code>, <code>genedata</code>, and <code>sequence</code>. The <code>name</code> is read from the <code>header</code>, which is stored as a string. The annotation data is stored in <code>genedata</code>, but generally you should use <code>genes</code> to access that data. For example, it can be used to iterate over annotations, and to modify them.</p><pre><code class="language-julia">for gene in chr.genes
    gene.locus_tag = &quot;$(chr.name)_$(gene.locus_tag)&quot;
end

chr.genes[2].locus_tag = &quot;test123&quot;</code></pre><p>The macro <code>@genes</code> can be used to filter through the annotations (see <a href="genes/#GenomicAnnotations.@genes"><code>@genes</code></a>). The keyword <code>gene</code> is used to refer to the individual <code>Gene</code>s. <code>@genes</code> can also be used to modify annotations.</p><pre><code class="language-julia">@genes(chr, length(gene) &gt; 300) # Returns all features longer than 300 nt</code></pre><p>Gene sequences can be accessed with <code>sequence(gene)</code>. For example, the following code will write the translated sequences of all protein-coding genes in <code>chr</code> to a file:</p><pre><code class="language-julia">using BioSequences
using FASTX
open(FASTA.Writer, &quot;proteins.fasta&quot;) do w
    for gene in @genes(chr, CDS)
        aaseq = translate(sequence(gene))
        write(w, FASTA.record(gene.locus_tag, get(:product, &quot;&quot;), aaseq))
    end
end</code></pre><p>Genes can be added using <code>addgene!</code>, and <code>sort!</code> can be used to make sure that the resulting annotations are in the correct order for printing. <code>delete!</code> is used to remove genes.</p><pre><code class="language-julia">newgene = addgene!(chr, &quot;regulatory&quot;, 670:677)
newgene.locus_tag = &quot;reg02&quot;
sort!(chr.genes)

# Genes can be deleted. This works for all genes where `:pseudo` is `true`, and ignores genes where it is `false` or `missing`
delete!(@genes(chr, :pseudo))
# Delete all genes 60 nt or shorter
delete!(@genes(chr, length(gene) &lt;= 60))</code></pre><p>Individual genes, and <code>Vector{Gene}</code>s are printed in GBK format. To include the GBK header and the nucleotide sequence, <code>printgbk(io, chr)</code> can be used to write them to a file.</p><pre><code class="language-julia">println(chr.genes[1])
println(@genes(chr, CDS))

open(&quot;updated.gbk&quot;, &quot;w&quot;) do f
    printgbk(f, chr)
end</code></pre><footer><hr/><a class="next" href="io/"><span class="direction">Next</span><span class="title">I/O</span></a></footer></article></body></html>
