<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>GenomicAnnotations.jl · GenomicAnnotations.jl</title><meta name="title" content="GenomicAnnotations.jl · GenomicAnnotations.jl"/><meta property="og:title" content="GenomicAnnotations.jl · GenomicAnnotations.jl"/><meta property="twitter:title" content="GenomicAnnotations.jl · GenomicAnnotations.jl"/><meta name="description" content="Documentation for GenomicAnnotations.jl."/><meta property="og:description" content="Documentation for GenomicAnnotations.jl."/><meta property="twitter:description" content="Documentation for GenomicAnnotations.jl."/><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="search_index.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href>GenomicAnnotations.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>GenomicAnnotations.jl</a><ul class="internal"><li><a class="tocitem" href="#Description"><span>Description</span></a></li><li><a class="tocitem" href="#Installation"><span>Installation</span></a></li><li><a class="tocitem" href="#Examples"><span>Examples</span></a></li></ul></li><li><a class="tocitem" href="io/">I/O</a></li><li><a class="tocitem" href="accessing/">Accessing and modifying annotations</a></li><li><a class="tocitem" href="loci/">Representing genomic loci</a></li><li><a class="tocitem" href="genes/">Filtering: the @genes macro</a></li><li><a class="tocitem" href="examples/">Examples</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>GenomicAnnotations.jl</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>GenomicAnnotations.jl</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/BioJulia/GenomicAnnotations.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/BioJulia/GenomicAnnotations.jl/blob/master/docs/src/index.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="GenomicAnnotations.jl"><a class="docs-heading-anchor" href="#GenomicAnnotations.jl">GenomicAnnotations.jl</a><a id="GenomicAnnotations.jl-1"></a><a class="docs-heading-anchor-permalink" href="#GenomicAnnotations.jl" title="Permalink"></a></h1><h2 id="Description"><a class="docs-heading-anchor" href="#Description">Description</a><a id="Description-1"></a><a class="docs-heading-anchor-permalink" href="#Description" title="Permalink"></a></h2><p>GenomicAnnotations is a package for reading, modifying, and writing genomic annotations in the GenBank and GFF3 file formats.</p><h2 id="Installation"><a class="docs-heading-anchor" href="#Installation">Installation</a><a id="Installation-1"></a><a class="docs-heading-anchor-permalink" href="#Installation" title="Permalink"></a></h2><pre><code class="language-julia hljs">julia&gt;]
pkg&gt; add GenomicAnnotations</code></pre><h2 id="Examples"><a class="docs-heading-anchor" href="#Examples">Examples</a><a id="Examples-1"></a><a class="docs-heading-anchor-permalink" href="#Examples" title="Permalink"></a></h2><p>GenBank and GFF3 files are read with <code>readgbk(input)</code> and <code>readgff(input)</code>, which return vectors of <code>Record</code>s. <code>input</code> can be an <code>IOStream</code> or a file path. GZipped data can be read by setting the keyword <code>gunzip</code> to true, which is done automatically if a filename ending in &quot;.gz&quot; is passed as <code>input</code>. If we&#39;re only interested in the first chromosome in <code>example.gbk</code> we only need to store the first record.</p><pre><code class="language-julia hljs">chr = readgbk(&quot;test/example.gbk&quot;)[1]</code></pre><p>Another way to read files is to use the corresponding <code>Reader</code> directly:</p><pre><code class="language-julia hljs">open(GenBank.Reader, &quot;test/example.gbk&quot;) do reader
    for record in reader
        println(record.name)
    end
end</code></pre><p><code>Record</code>s have five fields, <code>name</code>, <code>header</code>, <code>genes</code>, <code>genedata</code>, and <code>sequence</code>. The <code>name</code> is read from the <code>header</code>, which is stored as a string. The annotation data is stored in <code>genedata</code>, but generally you should use <code>genes</code> to access that data. For example, it can be used to iterate over annotations, and to modify them.</p><pre><code class="language-julia hljs">for gene in chr.genes
    gene.locus_tag = &quot;$(chr.name)_$(gene.locus_tag)&quot;
end

chr.genes[2].locus_tag = &quot;test123&quot;</code></pre><p>The locus of a <code>Gene</code> is represented by an <code>AbstractLocus</code> (see <a href="loci/#Loci">Loci</a>), which can be retrieved with <code>locus(gene)</code>. The locus of a gene can be updated with <code>locus!(gene, newlocus)</code>. The easiest way to create a locus is to use the constructor <code>Locus(s)</code>, which takes an <code>AbstractString</code> <code>s</code> and parses it as a GenBank locus string as defined here: https://www.insdc.org/submitting-standards/feature-table/#3.4. Note that remote entry descriptors have not been implemented.</p><pre><code class="language-julia hljs"># Creating a new locus
newlocus = Locus(&quot;complement(join(1..100,200..&gt;300))&quot;)

# Assigning a new locus to a gene
locus!(gene, newlocus)
# which is equivalent to
locus!(gene, &quot;complement(join(1..100,200..&gt;300))&quot;)</code></pre><p>For simplicity, <code>position(gene)</code> is shorthand for <code>locus(gene).position</code>. <code>locus(gene).position</code> gives an iteratable object that generates each individual position in the defined order. Thus:</p><pre><code class="language-julia hljs">loc = Locus(&quot;join(4..6,1..3)&quot;)
collect(loc.position) # Returns [4,5,6,1,2,3]</code></pre><p>The macro <code>@genes</code> can be used to filter through the annotations (see <a href="genes/#GenomicAnnotations.@genes"><code>@genes</code></a>). The keyword <code>gene</code> is used to refer to the individual <code>Gene</code>s. <code>@genes</code> can also be used to modify annotations.</p><pre><code class="language-julia hljs">@genes(chr, length(gene) &gt; 300) # Returns all features longer than 300 nt

@genes(chr, CDS, ismissing(:product)) .= &quot;hypothetical product&quot;</code></pre><p>Gene sequences can be accessed with <code>sequence(gene)</code>, which returns the nucleotide sequence. If the <code>translate</code> keyword is set to <code>true</code>, the translated amino acid sequence is returned instead. By default the first codon is translated to methionine also for alternate start codons, but this behaviour can be toggled by setting <code>preserve_alternate_start</code> to <code>false</code>. No checks are made to ensure that the gene points to a valid open reading frame, so this should be done by the user. The following example will write the translated sequences of all protein-coding genes in <code>chr</code> to a file:</p><pre><code class="language-julia hljs">using BioSequences
using FASTX
open(FASTA.Writer, &quot;proteins.fasta&quot;) do w
    for gene in @genes(chr, CDS, iscomplete(gene))
        aaseq = GenomicAnnotations.sequence(gene; translate = true)
        write(w, FASTA.Record(gene.locus_tag, get(:product, &quot;&quot;), aaseq))
    end
end</code></pre><p>Genes can be added using <code>addgene!</code>, and <code>sort!</code> can be used to make sure that the resulting annotations are in the correct order for printing. <code>delete!</code> is used to remove genes.</p><pre><code class="language-julia hljs">newgene = addgene!(chr, &quot;regulatory&quot;, 670:677)
newgene.locus_tag = &quot;reg02&quot;
sort!(chr.genes)

# Genes can be deleted. This works for all genes where `:pseudo` is `true`, and ignores genes where it is `false` or `missing`
delete!(@genes(chr, :pseudo))
# Delete all genes 60 nt or shorter
delete!(@genes(chr, length(gene) &lt;= 60))</code></pre><p>Individual genes, and <code>Vector{Gene}</code>s are printed in GBK format. To include the GBK header and the nucleotide sequence, <code>write(::GenBank.Writer, chr)</code> can be used to write them to a file. Use <code>GFF.Writer</code> instead to print the annotations as GFF3, in which case the GenBank header is lost.</p><pre><code class="language-julia hljs">println(chr.genes[1])
println(@genes(chr, CDS))

open(GenBank.Writer, &quot;updated.gbk&quot;) do w
    write(w, chr)
end</code></pre></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="io/">I/O »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Friday 22 November 2024 18:31">Friday 22 November 2024</span>. Using Julia version 1.11.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
