<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Accessing and modifying annotations · GenomicAnnotations.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>GenomicAnnotations.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../">GenomicAnnotations.jl</a></li><li><a class="toctext" href="../io/">I/O</a></li><li class="current"><a class="toctext" href>Accessing and modifying annotations</a><ul class="internal"><li class="toplevel"><a class="toctext" href="#Feature-1">Feature</a></li><li class="toplevel"><a class="toctext" href="#Qualifiers-1">Qualifiers</a></li><li class="toplevel"><a class="toctext" href="#Sequences-1">Sequences</a></li></ul></li><li><a class="toctext" href="../genes/">Filtering: the @genes macro</a></li></ul></nav><article id="docs"><header><nav><ul><li><a href>Accessing and modifying annotations</a></li></ul><a class="edit-page" href="https://github.com/kdyrhage/GenomicAnnotations.jl/blob/master/docs/src/accessing.md"><span class="fa"></span> Edit on GitHub</a></nav><hr/><div id="topbar"><span>Accessing and modifying annotations</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Accessing-and-modifying-annotations-1" href="#Accessing-and-modifying-annotations-1">Accessing and modifying annotations</a></h1><h1><a class="nav-anchor" id="Feature-1" href="#Feature-1">Feature</a></h1><p>Features (genes) can be added using <code>addgene!</code>. A feature must have a feature name and a locus (position), and can have any number of additional qualifiers associated with it (see next section).</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicAnnotations.addgene!" href="#GenomicAnnotations.addgene!"><code>GenomicAnnotations.addgene!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">addgene!(chr::Chromosome, feature, locus; kw...)</code></pre><p>Add gene to <code>chr</code>. <code>locus</code> can be a <code>Locus</code>, a UnitRange, or a StepRange (for decreasing ranges, which will be annotated on the complementary strand).</p><p><strong>Example</strong></p><pre><code class="language-julia">addgene!(chr, &quot;CDS&quot;, 1:756;
    locus_tag = &quot;gene0001&quot;,
    product = &quot;Chromosomal replication initiator protein dnaA&quot;)</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/kdyrhage/GenomicAnnotations.jl/blob/63289cea62aff832c3c7c2f6a48fe07caf4da475/src/types.jl#L68-L80">source</a></section><p>After adding a new feature, <code>sort!</code> can be used to make sure that the annotations are stored (and printed) in the order in which they occur on the chromosome:</p><pre><code class="language-julia">sort!(chr)</code></pre><p>Existing features can be removed using <code>delete!</code>:</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.delete!-Tuple{Gene}" href="#Base.delete!-Tuple{Gene}"><code>Base.delete!</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-julia">delete!(gene::Gene)</code></pre><p>Delete <code>gene</code> from <code>gene.parent</code>. Warning: does not work when broadcasted! Use <code>delete!(::AbstractVector{Gene}) instead.</code></p></div></div><a class="source-link" target="_blank" href="https://github.com/kdyrhage/GenomicAnnotations.jl/blob/63289cea62aff832c3c7c2f6a48fe07caf4da475/src/types.jl#L94-L99">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.delete!-Tuple{AbstractArray{Gene,1}}" href="#Base.delete!-Tuple{AbstractArray{Gene,1}}"><code>Base.delete!</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-julia">delete!(genes::AbstractArray{Gene, 1})</code></pre><p>Delete all genes in <code>genes</code> from <code>genes[1].parent</code>.</p><p><strong>Example</strong></p><pre><code class="language-julia">delete!(@genes(chr, length(gene) &lt;= 60))</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/kdyrhage/GenomicAnnotations.jl/blob/63289cea62aff832c3c7c2f6a48fe07caf4da475/src/types.jl#L108-L117">source</a></section><h1><a class="nav-anchor" id="Qualifiers-1" href="#Qualifiers-1">Qualifiers</a></h1><p>Features can have multiple qualifiers, which can be modified using Julia&#39;s property syntax:</p><pre><code class="language-julia"># Remove newspace from gene product descriptions
for gene in @genes(chr, iscds)
    replace!(gene.product, &#39;\n&#39; =&gt; &#39; &#39;)
end</code></pre><p>Properties also work on views of genes, typically generated using <code>@genes</code>:</p><pre><code class="language-julia">interestinggenes = readlines(&quot;/path/to/list/of/interesting/genes.txt&quot;)
@genes(chr, iscds, :locus_tag in interestinggenes).interesting .= true</code></pre><p>Sometimes features have multiple instances of the same qualifier, such genes having several EC-numbers. Assigning qualifiers with property syntax overwrites any data that was previously stored for that feature, and trying to assign a vector of values to a qualifier that is currently storing scalars will result in an error, so to safely assign qualifiers that might have more instances one can use <code>pushproperty!</code>:</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicAnnotations.pushproperty!" href="#GenomicAnnotations.pushproperty!"><code>GenomicAnnotations.pushproperty!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-julia">pushproperty!(gene::AbstractGene, qualifier::Symbol, value::T)</code></pre><p>Add a property to <code>gene</code>, similarly to <code>Base.setproperty!(::gene)</code>, but if the property is not missing in <code>gene</code>, it will be transformed to store a vector instead of overwriting existing data.</p><pre><code class="language-julia">julia&gt; eltype(chr.genedata[!, :EC_number])
Union{Missing,String}

julia&gt; chr.genes[1].EC_number = &quot;EC:1.2.3.4&quot;
&quot;EC:1.2.3.4&quot;

julia&gt; pushproperty!(chr.genes[1], :EC_number, &quot;EC:4.3.2.1&quot;); chr.genes[1].EC_number
2-element Array{String,1}:
 &quot;EC:1.2.3.4&quot;
 &quot;EC:4.3.2.1&quot;

julia&gt; eltype(chr.genedata[!, :EC_number])
Union{Missing, Array{String,1}}</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/kdyrhage/GenomicAnnotations.jl/blob/63289cea62aff832c3c7c2f6a48fe07caf4da475/src/types.jl#L150-L172">source</a></section><p>Accessing properties that haven&#39;t been stored will return missing. For this reason, it often makes more sense to use <code>get()</code> than to access the property directly.</p><pre><code class="language-julia"># chr.genes[2].pseudo returns missing, so this will throw an error
if chr.genes[2].pseudo
    println(&quot;Gene 2 is a pseudogene&quot;)
end

# ... but this works:
if get(chr.genes[2], :pseudo, false)
    println(&quot;Gene 2 is a pseudogene&quot;)
end</code></pre><h1><a class="nav-anchor" id="Sequences-1" href="#Sequences-1">Sequences</a></h1><p>The sequence of a <code>Chromosome</code> <code>chr</code> is stored in <code>chr.sequence</code>. Sequences of individual features can be read with <code>sequence</code>:</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicAnnotations.sequence-Tuple{Gene}" href="#GenomicAnnotations.sequence-Tuple{Gene}"><code>GenomicAnnotations.sequence</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-julia">sequence(gene::AbstractGene)</code></pre><p>Return genomic sequence for <code>gene</code>.</p><p><strong>Example</strong></p><p>The following code will write the translated sequences of all genes in <code>chr</code> to a file:</p><pre><code class="language-julia">using FASTX
writer = FASTA.Writer(open(&quot;proteins.fasta&quot;, &quot;w&quot;))
for gene in @genes(chr, iscds)
    aaseq = translate(sequence(gene))
    write(writer, FASTA.record(gene.locus_tag, get(gene, :product, &quot;&quot;), aaseq))
end
close(writer)</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/kdyrhage/GenomicAnnotations.jl/blob/63289cea62aff832c3c7c2f6a48fe07caf4da475/src/types.jl#L231-L247">source</a></section><footer><hr/><a class="previous" href="../io/"><span class="direction">Previous</span><span class="title">I/O</span></a><a class="next" href="../genes/"><span class="direction">Next</span><span class="title">Filtering: the @genes macro</span></a></footer></article></body></html>
