<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Filtering: the @genes macro · GenomicAnnotations.jl</title><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.11.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.11.1/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit">GenomicAnnotations.jl</span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">GenomicAnnotations.jl</a></li><li><a class="tocitem" href="../io/">I/O</a></li><li><a class="tocitem" href="../accessing/">Accessing and modifying annotations</a></li><li class="is-active"><a class="tocitem" href>Filtering: the @genes macro</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Filtering: the @genes macro</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Filtering: the @genes macro</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/kdyrhage/GenomicAnnotations.jl/blob/master/docs/src/genes.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Filtering:-the-@genes-macro-1"><a class="docs-heading-anchor" href="#Filtering:-the-@genes-macro-1">Filtering: the @genes macro</a><a class="docs-heading-anchor-permalink" href="#Filtering:-the-@genes-macro-1" title="Permalink"></a></h1><p>A useful tool provided by GenomicAnnotations is the macro <code>@genes</code>. It is used to filter through annotations, for example to look at only at coding sequences or rRNAs, which can then be modified or iterated over:</p><pre><code class="language-julia"># Print locus tags of all coding sequences longer than 1000 nt, that are not pseudo genes
for gene in @genes(chr, CDS, length(gene) &gt; 1000, ! :pseudo)
    println(gene.locus_tag)
end</code></pre><article class="docstring"><header><a class="docstring-binding" id="GenomicAnnotations.@genes" href="#GenomicAnnotations.@genes"><code>GenomicAnnotations.@genes</code></a> — <span class="docstring-category">Macro</span></header><section><div><pre><code class="language-julia">@genes(chr, exs...)</code></pre><p>Iterate over and evaluate expressions in <code>exs</code> for all genes in <code>chr.genes</code>, returning genes where all expressions evaluate to <code>true</code>. Any given symbol <code>s</code> in the expression will be substituted for <code>gene.s</code>. The gene itself can be accessed in the expression as <code>gene</code>. Accessing properties of the returned list of genes returns a view, which can be altered.</p><p>Symbols and expressions escaped with <code>Ref()</code> will be ignored.</p><p>Some short-hand forms are available to make life easier:     <code>CDS</code>, <code>rRNA</code>, and <code>tRNA</code> expand to <code>feature(gene) == &quot;...&quot;</code>,     <code>get(s::Symbol, default)</code> expands to <code>get(gene, s, default)</code></p><p><strong>Examples</strong></p><pre><code class="language-julia">julia&gt; chromosome = readgbk(&quot;example.gbk&quot;)
Chromosome &#39;example&#39; (5028 bp) with 6 annotations

julia&gt; @genes(chromosome, iscds) |&gt; length
3

julia&gt; @genes(chromosome, length(gene) &lt; 500)
     CDS             3..206
                     /db_xref=&quot;GI:1293614&quot;
                     /locus_tag=&quot;tag01&quot;
                     /codon_start=&quot;3&quot;
                     /product=&quot;TCP1-beta&quot;
                     /protein_id=&quot;AAA98665.1&quot;

julia&gt; @genes(chromosome, ismissing(:gene)) |&gt; length
2

julia&gt; @genes(chromosome, ismissing(:gene)).gene .= &quot;Unknown&quot;;

julia&gt; @genes(chromosome, ismissing(:gene)) |&gt; length
0</code></pre><p>All arguments have to evaluate to <code>true</code> for a gene to be included, so the following expressions are equivalent:</p><pre><code class="language-julia">@genes(chr, feature(gene) == Ref(:CDS), length(gene) &gt; 300)
@genes(chr, (feature(gene) == Ref(:CDS)) &amp;&amp; (length(gene) &gt; 300))</code></pre><p><code>@genes</code> returns a <code>Vector{Gene}</code>. Attributes can be accessed with dot-syntax, and can be assigned to</p><pre><code class="language-julia">@genes(chr, :locus_tag == &quot;tag03&quot;)[1].pseudo = true
@genes(chr, CDS, ismissing(:gene)).gene .= &quot;unknown&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/kdyrhage/GenomicAnnotations.jl/blob/e6c35dcf01ef4da6c7d0c8114fe8413d3907f05b/src/macro.jl#L44-L95">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../accessing/">« Accessing and modifying annotations</a></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> on <span class="colophon-date" title="Friday 13 December 2019 13:04">Friday 13 December 2019</span>. Using Julia version 1.3.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
